static_analyzer/program_info/source_facts.py

```
@@ -0,0 +1,791 @@
+        tuple(sorted(set(types))),
+        tuple(sorted(set(annotations))),
+        tuple(sorted(set(declarations))),
+        tuple(sorted(set(diagnostics))),
+    )
+
+
+def enrich_and_resolve_source_facts(
+    facts: tuple[FileSourceFacts, ...], symbols: Mapping[str, object]
+) -> ResolvedSourceFacts:
+    """Attach facts to narrow symbols and resolve only unique structural targets."""
+    primary = list({getattr(symbol, "qualified_name"): symbol for symbol in symbols.values()}.values())
```

Ruff (0.16.0)

[warning] Do not call `getattr` with a constant attribute value. It is not any safer than normal property access.

Replace `getattr` with attribute access

(B009)